NumLib/ODESolver/NonlinearSolver.cpp

```diff
@@ -215,63 +215,65 @@ bool NonlinearSolver<NonlinearSolverTag::Newton>::solve(
         sys.applyKnownSolutionsNewton(J, res, minus_delta_x);
         INFO("[time] Applying Dirichlet BCs took %g s.", time_dirichlet.elapsed());
 
-        if (!sys.isLinear() && _convergence_criterion->hasResidualCheck())
+        if (_convergence_criterion->hasResidualCheck())
             _convergence_criterion->checkResidual(res);
 
-        BaseLib::RunTime time_linear_solver;
-        time_linear_solver.start();
-        bool iteration_succeeded = _linear_solver.solve(J, res, minus_delta_x);
-        INFO("[time] Linear solver took %g s.", time_linear_solver.elapsed());
-
-        if (!iteration_succeeded)
+        GlobalVector* x_old = nullptr;
+        if (_convergence_criterion->hasResidualCheck() && _convergence_criterion->isSatisfied())
```

Comment:
it's because, if residual is enough small with current solution, we don't need to solve a linear system. If you use a convergence criteria based on delta x, convergence cannot be checked until you get next solution, i.e. solve a linear system.